sched/semaphore/sem_holder.c

```diff
@@ -566,6 +566,8 @@ void nxsem_initialize_holders(void)
 
 void nxsem_destroyholder(FAR sem_t *sem)
 {
+  irqstate_t flags = enter_critical_section();
```

Comment:
but if other place doesn't hold the critical section, how to avoid them modifying the list?

Reply:
Every other call that modifies the list comes from within critical section, this is the only hole currently.

Reply:
If so, should we add the lock in caller site, like the approach taken by other function

Reply:
We can do it like that as well

Reply:
since other funtion in this file assume the caller hold the critical section, nxsem_destroyholder is better to follow this direction too, otherwise.

Reply:
@xiaoxiang781216 I was looking into doing the change and I noticed that it will make the caller's side very messy. nxsem_destroyholder is defined as empty macro if priority inheritance is not enabled:
https://github.com/apache/nuttx/blob/5d68ab635c065d97ac0cbc08372be7c37fb59d02/sched/semaphore/semaphore.h#L81-L91

The callers that do not run inside critical section are as follows:
https://github.com/apache/nuttx/blob/5d68ab635c065d97ac0cbc08372be7c37fb59d02/sched/semaphore/sem_destroy.c#L61-L91

https://github.com/apache/nuttx/blob/5d68ab635c065d97ac0cbc08372be7c37fb59d02/sched/semaphore/sem_setprotocol.c#L77-L109

If the critical section is moved to the caller side, it will be quite ugly because we need to flag the critical section out if priority inheritance is off:

```
int nxsem_set_protocol(FAR sem_t *sem, int protocol)
{
  DEBUGASSERT(sem != NULL);

#ifdef CONFIG_PRIORITY_INHERITANCE
  irqstate_t flags;
#endif

  switch (protocol & SEM_PRIO_MASK)
    {
      case SEM_PRIO_NONE:

        /* Remove any current holders */

#ifdef CONFIG_PRIORITY_INHERITANCE
        flags = enter_critical_section();
#endif
        nxsem_destroyholder(sem);
#ifdef CONFIG_PRIORITY_INHERITANCE
        leave_critical_section(flags);
#endif
        break;

      case SEM_PRIO_INHERIT:

        break;

      case SEM_PRIO_PROTECT:
#ifdef CONFIG_PRIORITY_PROTECT
        break;
#else
        /* Not yet supported */

        return -ENOTSUP;
#endif

      default:
        return -EINVAL;
    }

  sem->flags = protocol;
  return OK;
}

int nxsem_destroy(FAR sem_t *sem)
{
  int32_t old;
#ifdef CONFIG_PRIORITY_INHERITANCE
  irqstate_t flags;
#endif

  DEBUGASSERT(sem != NULL);

  /* There is really no particular action that we need
   * take to destroy a semaphore.  We will just reset
   * the count to some reasonable value (0) and release
   * ownership.
   *
   * Check if other threads are waiting on the semaphore.
   * In this case, the behavior is undefined.  We will:
   * leave the count unchanged but still return OK.
   */

  old = atomic_read(NXSEM_COUNT(sem));
  do
    {
      if (old < 0)
        {
          break;
        }
    }
  while (!atomic_try_cmpxchg_release(NXSEM_COUNT(sem), &old, 1));

  /* Release holders of the semaphore */

#ifdef CONFIG_PRIORITY_INHERITANCE
  flags = enter_critical_section();
#endif
  nxsem_destroyholder(sem);
#ifdef CONFIG_PRIORITY_INHERITANCE
  leave_critical_section(flags);
#endif
  return OK;
}
```

Maybe some of the #ifdef areas could be merged but it will still be pretty ugly IMO. I suggest we keep this patch as is. What do you think ?

Reply:
Ok, let's keep the current form